File: combinatorics/resource/LCHP/problem_LTDT_code/bai_13/bai13.py
from collections import defaultdict
from typing import List, Dict, Set, Tuple, Optional
import copy

class Edge:
    """Lớp biểu diễn cạnh trong đồ thị tổng quát"""
    
    def __init__(self, to: int, is_directed: bool, edge_id: int):
        self.to = to
        self.is_directed = is_directed
        self.id = edge_id

class EdgeInfo:
    """Thông tin chi tiết về cạnh"""
    
    def __init__(self, from_vertex: int, to_vertex: int, edge_id: int, 
                 is_directed: bool, edge_type: str, is_self_loop: bool):
        self.from_vertex = from_vertex
        self.to_vertex = to_vertex
        self.id = edge_id
        self.is_directed = is_directed
        self.type = edge_type
        self.is_self_loop = is_self_loop

class GeneralGraphDFS:
    """DFS cho đồ thị tổng quát"""
    
    def __init__(self, num_vertices: int):
        self.num_vertices = num_vertices
        self.adj_list = [[] for _ in range(num_vertices)]
        self.edge_map = defaultdict(list)
        self.self_loop_count = [0] * num_vertices
        self.edge_id_counter = 0
        self.reset_arrays()
    
    def reset_arrays(self):
        self.visited = [False] * self.num_vertices
        self.start_time = [0] * self.num_vertices
        self.finish_time = [0] * self.num_vertices
        self.parent = [-1] * self.num_vertices
        self.dfs_order = []
        self.edge_info_list = []
        self.time_counter = 0
    
    def add_edge(self, u: int, v: int, is_directed: bool = False):
        """Thêm cạnh vào đồ thị tổng quát"""
        edge = Edge(v, is_directed, self.edge_id_counter)
        self.edge_id_counter += 1
        self.adj_list[u].append(edge)
        
        # Xử lý khuyên
        if u == v:
            self.self_loop_count[u] += 1
            print(f"Thêm khuyên tại đỉnh {u} (ID: {edge.id})")
        else:
            # Cạnh thường - thêm vào cả 2 chiều nếu vô hướng
            if not is_directed:
                reverse_edge = Edge(u, False, self.edge_id_counter)
                self.edge_id_counter += 1
                self.adj_list[v].append(reverse_edge)
            
            direction_str = "có hướng" if is_directed else "vô hướng"
            print(f"Thêm cạnh {u} -> {v} ({direction_str}) (ID: {edge.id})")
        
        # Cập nhật bản đồ cạnh
        edge_key = (u, v)
        self.edge_map[edge_key].append(edge)
    
    def dfs_recursive(self, u: int):
        """DFS đệ quy cho đồ thị tổng quát"""
        self.visited[u] = True
        self.time_counter += 1
        self.start_time[u] = self.time_counter
        self.dfs_order.append(u)
        
        print(f"Thăm đỉnh {u} (start: {self.start_time[u]})")
        
        # Xử lý khuyên trước
        if self.self_loop_count[u] > 0:
            print(f"  Phát hiện {self.self_loop_count[u]} khuyên tại đỉnh {u}")
            self.edge_info_list.append(
                EdgeInfo(u, u, -1, True, "SELF_LOOP", True)
            )
        
        # Duyệt các cạnh thường
        for edge in self.adj_list[u]:
            v = edge.to
            is_directed = edge.is_directed
            
            # Bỏ qua khuyên (đã xử lý ở trên)
            if u == v:
                continue
            
            if not self.visited[v]:
                self.parent[v] = u
                self.edge_info_list.append(
                    EdgeInfo(u, v, edge.id, is_directed, "TREE", False)
                )
                direction_str = "có hướng" if is_directed else "vô hướng"
                print(f"  Cạnh cây: {u} -> {v} ({direction_str})")
                self.dfs_recursive(v)
            else:
                # Phân loại cạnh phức tạp
                edge_type = self._classify_edge(u, v, is_directed)
                self.edge_info_list.append(
                    EdgeInfo(u, v, edge.id, is_directed, edge_type, False)
                )
                
                direction_str = "có hướng" if is_directed else "vô hướng"
                print(f"  Cạnh {edge_type}: {u} -> {v} ({direction_str})", end="")
                
                if edge_type == "BACK":
                    print(" (chu trình!)", end="")
                print()
        
        self.time_counter += 1
        self.finish_time[u] = self.time_counter
        print(f"Kết thúc đỉnh {u} (finish: {self.finish_time[u]})")
    
    def _classify_edge(self, u: int, v: int, is_directed: bool) -> str:
        """Phân loại cạnh dựa trên thời gian và hướng"""
        if (self.start_time[v] < self.start_time[u] and 
            self.finish_time[v] == 0):
            return "BACK"  # Cạnh ngược - tạo chu trình
        
        if self.start_time[v] > self.start_time[u]:
            return "FORWARD"  # Cạnh xuôi
        
        if self.finish_time[v] < self.start_time[u]:
            return "CROSS"  # Cạnh chéo
        
        # Trường hợp đặc biệt cho cạnh vô hướng
        if not is_directed and self.parent[u] == v:
            return "PARENT"  # Cạnh về cha (không tính chu trình)
        
        return "UNKNOWN"
    
    def dfs_complete(self):
        """DFS toàn bộ đồ thị tổng quát"""
        self.reset_arrays()
        
        print("DFS toàn bộ đồ thị tổng quát:")
        component_count = 0
        
        for i in range(self.num_vertices):
            if not self.visited[i]:
                component_count += 1
                print(f"\nThành phần liên thông {component_count}:")
                self.dfs_recursive(i)
        
        print(f"\nTổng số thành phần liên thông: {component_count}")
        self.analyze_complex_graph()
    
    def analyze_complex_graph(self):
        """Phân tích đồ thị tổng quát"""
        print("\nPhân tích đồ thị tổng quát:")
        
        tree_edges = back_edges = forward_edges = cross_edges = 0
        self_loops = directed_edges = undirected_edges = 0
        has_cycle = False
        
        for info in self.edge_info_list:
            if info.is_self_loop:
                self_loops += 1
                continue
            
            if info.is_directed:
                directed_edges += 1
            else:
                undirected_edges += 1
            
            if info.type == "TREE":
                tree_edges += 1
            elif info.type == "BACK":
                back_edges += 1
                has_cycle = True
            elif info.type == "FORWARD":
                forward_edges += 1
            elif info.type == "CROSS":
                cross_edges += 1
        
        print("Thống kê cạnh:")
        print(f"- Cạnh cây: {tree_edges}")
        print(f"- Cạnh ngược: {back_edges}")
        print(f"- Cạnh xuôi: {forward_edges}")
        print(f"- Cạnh chéo: {cross_edges}")
        print(f"- Khuyên: {self_loops}")
        print(f"- Cạnh có hướng: {directed_edges}")
        print(f"- Cạnh vô hướng: {undirected_edges}")
        
        print(f"\nĐồ thị có chu trình: {'Có' if has_cycle else 'Không'}")
        
        # Phân tích tính chất đặc biệt
        self.analyze_special_properties()
    
    def analyze_special_properties(self):
        """Phân tích tính chất đặc biệt"""
        print("\nPhân tích tính chất đặc biệt:")
        
        # Kiểm tra đồ thị Euler
        is_eulerian = self.check_eulerian()
        print(f"Đồ thị Euler: {'Có' if is_eulerian else 'Không'}")
        
        # Kiểm tra tính liên thông mạnh (cho phần có hướng)
        is_strongly_connected = self.check_strong_connectivity()
        print(f"Liên thông mạnh: {'Có' if is_strongly_connected else 'Không'}")
        
        # Đếm số bậc của mỗi đỉnh
        self.print_degree_information()
    
    def check_eulerian(self) -> bool:
        """Kiểm tra tính Euler đơn giản"""
        for i in range(self.num_vertices):
            degree = len(self.adj_list[i]) + self.self_loop_count[i]
            if degree % 2 != 0:
                return False
        return True
    
    def check_strong_connectivity(self) -> bool:
        """Kiểm tra tính liên thông mạnh đơn giản"""
        self.reset_arrays()
        self.dfs_recursive(0)
        
        for i in range(self.num_vertices):
            if not self.visited[i]:
                return False
        return True
    
    def print_degree_information(self):
        """In thông tin bậc đỉnh"""
        print("\nThông tin bậc đỉnh:")
        print("Đỉnh\tBậc vào\tBậc ra\tKhuyên")
        print("----\t-------\t------\t------")
        
        for i in range(self.num_vertices):
            in_degree = out_degree = 0
            
            # Tính bậc ra
            out_degree = len(self.adj_list[i])
            
            # Tính bậc vào
            for j in range(self.num_vertices):
                for edge in self.adj_list[j]:
                    if edge.to == i and j != i:
                        in_degree += 1
            
            print(f"{i}\t{in_degree}\t{out_degree}\t{self.self_loop_count[i]}")
    
    def print_general_graph(self):
        """In biểu diễn đồ thị tổng quát"""
        print("Biểu diễn đồ thị tổng quát:")
        
        for i in range(self.num_vertices):
            print(f"Đỉnh {i}:")
            
            if self.self_loop_count[i] > 0:
                print(f"  Khuyên: {self.self_loop_count[i]}")
            
            print("  Kề: ", end="")
            for edge in self.adj_list[i]:
                if edge.to != i:  # Không in khuyên ở đây
                    direction = "(→)" if edge.is_directed else "(↔)"
                    print(f"{edge.to}{direction} ", end="")
            print()
        print()
    
    def find_all_cycles(self):
        """Tìm tất cả chu trình trong đồ thị tổng quát"""
        print("Tìm tất cả chu trình trong đồ thị:")
        
        cycles = []
        current_path = []
        in_path = [False] * self.num_vertices
        
        for i in range(self.num_vertices):
            self._find_cycles_from_vertex(i, current_path, in_path, cycles)
        
        if not cycles:
            print("Không có chu trình")
        else:
            print(f"Tìm thấy {len(cycles)} chu trình:")
            for i, cycle in enumerate(cycles, 1):
                cycle_str = ' -> '.join(map(str, cycle))
                print(f"Chu trình {i}: {cycle_str} -> {cycle[0]}")
    
    def _find_cycles_from_vertex(self, start: int, path: List[int], 
                                in_path: List[bool], cycles: List[List[int]]):
        """Helper function để tìm chu trình từ một đỉnh"""
        path.append(start)
        in_path[start] = True
        
        for edge in self.adj_list[start]:
            next_vertex = edge.to
            
            if next_vertex == start:  # Bỏ qua khuyên
                continue
            
            if in_path[next_vertex]:
                # Tìm thấy chu trình
                try:
                    cycle_start = path.index(next_vertex)
                    cycle = path[cycle_start:]
                    cycles.append(cycle.copy())
                except ValueError:
                    pass
            elif next_vertex not in path:
                self._find_cycles_from_vertex(next_vertex, path, in_path, cycles)
        
        path.pop()
        in_path[start] = False
    
    def dfs_iterative(self, start: int):
        """DFS không đệ quy cho đồ thị tổng quát"""
        self.reset_arrays()
        
        stack = [start]
        print(f"DFS không đệ quy từ đỉnh {start}:")
        
        while stack:
            u = stack.pop()
            
            if not self.visited[u]:
                self.visited[u] = True
                self.time_counter += 1
                self.start_time[u] = self.time_counter
                self.dfs_order.append(u)
                
                print(f"Thăm đỉnh {u} (start: {self.start_time[u]})")
                
                # Xử lý khuyên
                if self.self_loop_count[u] > 0:
                    print(f"  Phát hiện {self.self_loop_count[u]} khuyên tại đỉnh {u}")
                
                # Thêm các đỉnh kề vào stack (thứ tự ngược lại)
                neighbors = []
                for edge in self.adj_list[u]:
                    if edge.to != u and not self.visited[edge.to]:
                        neighbors.append(edge.to)
                
                # Sắp xếp theo thứ tự ngược để có thứ tự duyệt đúng
                neighbors.sort(reverse=True)
                for v in neighbors:
                    stack.append(v)
                    if self.parent[v] == -1:
                        self.parent[v] = u
        print()
    
    def find_bridges_and_cut_vertices(self):
        """Tìm cầu và khớp trong đồ thị"""
        self.reset_arrays()
        
        low = [0] * self.num_vertices  # Giá trị low của mỗi đỉnh
        bridges = []  # Danh sách cầu
        cut_vertices = set()  # Tập hợp các đỉnh khớp
        
        def tarjan_bridge(u: int, parent: int):
            self.visited[u] = True
            self.time_counter += 1
            self.start_time[u] = low[u] = self.time_counter
            children = 0
            
            for edge in self.adj_list[u]:
                v = edge.to
                if u == v:  # Bỏ qua khuyên
                    continue
                
                if not self.visited[v]:
                    children += 1
                    tarjan_bridge(v, u)
                    
                    # Cập nhật low[u]
                    low[u] = min(low[u], low[v])
                    
                    # Kiểm tra cầu
                    if low[v] > self.start_time[u]:
                        bridges.append((u, v))
                    
                    # Kiểm tra đỉnh khớp
                    if (parent == -1 and children > 1) or (parent != -1 and low[v] >= self.start_time[u]):
                        cut_vertices.add(u)
                        
                elif v != parent:
                    low[u] = min(low[u], self.start_time[v])
        
        # Chạy Tarjan cho từng thành phần liên thông
        for i in range(self.num_vertices):
            if not self.visited[i]:
                tarjan_bridge(i, -1)
        
        print("Kết quả tìm cầu và khớp:")
        print(f"Số cầu: {len(bridges)}")
        if bridges:
            print("Danh sách cầu:")
            for bridge in bridges:
                print(f"  ({bridge[0]}, {bridge[1]})")
        
        print(f"Số đỉnh khớp: {len(cut_vertices)}")
        if cut_vertices:
            print("Danh sách đỉnh khớp:", sorted(cut_vertices))
    
    def topological_sort(self):
        """Sắp xếp topo (chỉ có ý nghĩa với đồ thị có hướng không có chu trình)"""
        self.reset_arrays()
        topo_stack = []
        
        def topo_dfs(u: int):
            self.visited[u] = True
            
            for edge in self.adj_list[u]:
                v = edge.to
                if not self.visited[v] and edge.is_directed:  # Chỉ theo cạnh có hướng
                    topo_dfs(v)
            
            topo_stack.append(u)
        
        # Gọi DFS cho tất cả đỉnh
        for i in range(self.num_vertices):
            if not self.visited[i]:
                topo_dfs(i)
        
        result = topo_stack[::-1]
        
        print("Sắp xếp topo (chỉ xét cạnh có hướng):")
        print("Thứ tự:", ' -> '.join(map(str, result)))
        
        return result
    
    def print_times(self):
        """In thông tin thời gian DFS"""
        print("\nThông tin thời gian DFS:")
        print(f"{'Đỉnh':>6} {'Start':>12} {'Finish':>12} {'Cha':>10}")
        print("-" * 40)
        
        for i in range(self.num_vertices):
            print(f"{i:>6} {self.start_time[i]:>12} {self.finish_time[i]:>12} {self.parent[i]:>10}")
        print()
    
    def print_dfs_tree(self):
        """In cây DFS"""
        print("Cây DFS:")
        for i in range(self.num_vertices):
            if self.parent[i] != -1:
                print(f"{self.parent[i]} -> {i}")
        print()

# Demo cho đồ thị tổng quát
def demonstrate_general_graph_dfs():
    print("=== DEMO DFS CHO ĐỒ THỊ TỔNG QUÁT ===\n")
    
    graph = GeneralGraphDFS(6)
    
    # Tạo đồ thị tổng quát phức tạp
    print("Tạo đồ thị tổng quát:")
    
    # Cạnh vô hướng
    graph.add_edge(0, 1, False)
    graph.add_edge(1, 2, False)
    
    # Cạnh có hướng
    graph.add_edge(2, 3, True)
    graph.add_edge(3, 4, True)
    
    # Cạnh bội
    graph.add_edge(0, 1, False)  # Cạnh bội vô hướng
    graph.add_edge(2, 3, True)   # Cạnh bội có hướng
    
    # Khuyên
    graph.add_edge(4, 4, True)   # Khuyên có hướng
    graph.add_edge(5, 5, False)  # Khuyên vô hướng
    
    # Cạnh tạo chu trình
    graph.add_edge(4, 1, True)   # Chu trình có hướng
    graph.add_edge(3, 0, False)  # Chu trình hỗn hợp
    
    print()
    graph.print_general_graph()
    
    # DFS trên đồ thị tổng quát
    print("1. DFS ĐỆ QUY:")
    graph.dfs_complete()
    graph.print_times()
    graph.print_dfs_tree()
    
    # DFS không đệ quy
    print("2. DFS KHÔNG ĐỆ QUY:")
    graph.dfs_iterative(0)
    
    # Tìm chu trình
    print("3. TÌM CHU TRÌNH:")
    graph.find_all_cycles()
    
    # Tìm cầu và khớp
    print("4. TÌM CẦU VÀ KHỚP:")
    graph.find_bridges_and_cut_vertices()
    
    # Sắp xếp topo
    print("5. SẮP XẾP TOPO:")
    graph.topological_sort()

if __name__ == "__main__":
    demonstrate_general_graph_dfs()

File: combinatorics/resource/LCHP/problem_LTDT_code/bai_13/bai13.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <unordered_map>
#include <algorithm>

using namespace std;

struct Edge {
    int to;
    bool isDirected;  // true = có hướng, false = vô hướng
    int id;          // ID duy nhất của cạnh
    
    Edge(int t, bool dir, int i) : to(t), isDirected(dir), id(i) {}
};

class GeneralGraphDFS {
private:
    int numVertices;
    vector<vector<Edge>> adjList;
    map<pair<int,int>, vector<Edge>> edgeMap;  // (u,v) -> list of edges
    
    vector<bool> visited;
    vector<int> startTime;
    vector<int> finishTime;
    vector<int> parent;
    vector<int> dfsOrder;
    int timeCounter;
    int edgeIdCounter;
    
    // Thông tin chi tiết về cạnh
    struct EdgeInfo {
        int from, to, id;
        bool isDirected;
        string type;
        bool isSelfLoop;
        
        EdgeInfo(int f, int t, int i, bool dir, string typ, bool self) 
            : from(f), to(t), id(i), isDirected(dir), type(typ), isSelfLoop(self) {}
    };
    
    vector<EdgeInfo> edgeInfoList;
    
    // Đếm số khuyên tại mỗi đỉnh
    vector<int> selfLoopCount;
    
public:
    GeneralGraphDFS(int n) : numVertices(n), timeCounter(0), edgeIdCounter(0) {
        adjList.resize(n);
        selfLoopCount.resize(n, 0);
        resetArrays();
    }
    
    void resetArrays() {
        visited.assign(numVertices, false);
        startTime.assign(numVertices, 0);
        finishTime.assign(numVertices, 0);
        parent.assign(numVertices, -1);
        dfsOrder.clear();
        edgeInfoList.clear();
        timeCounter = 0;
    }
    
    void addEdge(int u, int v, bool isDirected = false) {
        Edge edge(v, isDirected, edgeIdCounter++);
        adjList[u].push_back(edge);
        
        // Xử lý khuyên
        if (u == v) {
            selfLoopCount[u]++;
            cout << "Thêm khuyên tại đỉnh " << u << " (ID: " << edge.id << ")" << endl;
        } else {
            // Cạnh thường - thêm vào cả 2 chiều nếu vô hướng
            if (!isDirected) {
                Edge reverseEdge(u, false, edgeIdCounter++);
                adjList[v].push_back(reverseEdge);
            }
            
            cout << "Thêm cạnh " << u << " -> " << v 
                 << (isDirected ? " (có hướng)" : " (vô hướng)") 
                 << " (ID: " << edge.id << ")" << endl;
        }
        
        // Cập nhật bản đồ cạnh
        pair<int,int> edgeKey = {u, v};
        edgeMap[edgeKey].push_back(edge);
    }
    
    // DFS cho đồ thị tổng quát
    void dfsRecursive(int u) {
        visited[u] = true;
        startTime[u] = ++timeCounter;
        dfsOrder.push_back(u);
        
        cout << "Thăm đỉnh " << u << " (start: " << startTime[u] << ")" << endl;
        
        // Xử lý khuyên trước
        if (selfLoopCount[u] > 0) {
            cout << "  Phát hiện " << selfLoopCount[u] << " khuyên tại đỉnh " << u << endl;
            // Khuyên luôn được coi là cạnh đặc biệt
            edgeInfoList.push_back(EdgeInfo(u, u, -1, true, "SELF_LOOP", true));
        }
        
        // Duyệt các cạnh thường
        for (const Edge& edge : adjList[u]) {
            int v = edge.to;
            bool isDirected = edge.isDirected;
            
            // Bỏ qua khuyên (đã xử lý ở trên)
            if (u == v) continue;
            
            if (!visited[v]) {
                parent[v] = u;
                edgeInfoList.push_back(EdgeInfo(u, v, edge.id, isDirected, "TREE", false));
                cout << "  Cạnh cây: " << u << " -> " << v 
                     << (isDirected ? " (có hướng)" : " (vô hướng)") << endl;
                dfsRecursive(v);
            } else {
                // Phân loại cạnh phức tạp
                string edgeType = classifyEdge(u, v, isDirected);
                edgeInfoList.push_back(EdgeInfo(u, v, edge.id, isDirected, edgeType, false));
                
                cout << "  Cạnh " << edgeType << ": " << u << " -> " << v 
                     << (isDirected ? " (có hướng)" : " (vô hướng)");
                
                if (edgeType == "BACK") {
                    cout << " (chu trình!)";
                }
                cout << endl;
            }
        }
        
        finishTime[u] = ++timeCounter;
        cout << "Kết thúc đỉnh " << u << " (finish: " << finishTime[u] << ")" << endl;
    }
    
private:
    string classifyEdge(int u, int v, bool isDirected) {
        // Phân loại cạnh dựa trên thời gian và hướng
        if (startTime[v] < startTime[u] && finishTime[v] == 0) {
            return "BACK";  // Cạnh ngược - tạo chu trình
        }
        
        if (startTime[v] > startTime[u]) {
            return "FORWARD";  // Cạnh xuôi
        }
        
        if (finishTime[v] < startTime[u]) {
            return "CROSS";  // Cạnh chéo
        }
        
        // Trường hợp đặc biệt cho cạnh vô hướng
        if (!isDirected && parent[u] == v) {
            return "PARENT";  // Cạnh về cha (không tính chu trình)
        }
        
        return "UNKNOWN";
    }
    
public:
    void dfsComplete() {
        resetArrays();
        
        cout << "DFS toàn bộ đồ thị tổng quát:" << endl;
        int componentCount = 0;
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                componentCount++;
                cout << "\nThành phần liên thông " << componentCount << ":" << endl;
                dfsRecursive(i);
            }
        }
        
        cout << "\nTổng số thành phần liên thông: " << componentCount << endl;
        analyzeComplexGraph();
    }
    
    void analyzeComplexGraph() {
        cout << "\nPhân tích đồ thị tổng quát:" << endl;
        
        int treeEdges = 0, backEdges = 0, forwardEdges = 0, crossEdges = 0;
        int selfLoops = 0, directedEdges = 0, undirectedEdges = 0;
        bool hasCycle = false;
        
        for (const EdgeInfo& info : edgeInfoList) {
            if (info.isSelfLoop) {
                selfLoops++;
                continue;
            }
            
            if (info.isDirected) {
                directedEdges++;
            } else {
                undirectedEdges++;
            }
            
            if (info.type == "TREE") treeEdges++;
            else if (info.type == "BACK") { backEdges++; hasCycle = true; }
            else if (info.type == "FORWARD") forwardEdges++;
            else if (info.type == "CROSS") crossEdges++;
        }
        
        cout << "Thống kê cạnh:" << endl;
        cout << "- Cạnh cây: " << treeEdges << endl;
        cout << "- Cạnh ngược: " << backEdges << endl;
        cout << "- Cạnh xuôi: " << forwardEdges << endl;
        cout << "- Cạnh chéo: " << crossEdges << endl;
        cout << "- Khuyên: " << selfLoops << endl;
        cout << "- Cạnh có hướng: " << directedEdges << endl;
        cout << "- Cạnh vô hướng: " << undirectedEdges << endl;
        
        cout << "\nĐồ thị có chu trình: " << (hasCycle ? "Có" : "Không") << endl;
        
        // Phân tích tính chất đặc biệt
        analyzeSpecialProperties();
    }
    
    void analyzeSpecialProperties() {
        cout << "\nPhân tích tính chất đặc biệt:" << endl;
        
        // Kiểm tra đồ thị Euler
        bool isEulerian = checkEulerian();
        cout << "Đồ thị Euler: " << (isEulerian ? "Có" : "Không") << endl;
        
        // Kiểm tra tính liên thông mạnh (cho phần có hướng)
        bool isStronglyConnected = checkStrongConnectivity();
        cout << "Liên thông mạnh: " << (isStronglyConnected ? "Có" : "Không") << endl;
        
        // Đếm số bậc của mỗi đỉnh
        printDegreeInformation();
    }
    
    bool checkEulerian() {
        // Đơn giản hóa: kiểm tra bậc chẵn cho tất cả đỉnh
        for (int i = 0; i < numVertices; i++) {
            int degree = adjList[i].size() + selfLoopCount[i];
            if (degree % 2 != 0) {
                return false;
            }
        }
        return true;
    }
    
    bool checkStrongConnectivity() {
        // Kiểm tra đơn giản: tất cả đỉnh có thể đến được từ đỉnh 0
        resetArrays();
        dfsRecursive(0);
        
        for (int i = 0; i < numVertices; i++) {
            if (!visited[i]) {
                return false;
            }
        }
        return true;
    }
    
    void printDegreeInformation() {
        cout << "\nThông tin bậc đỉnh:" << endl;
        cout << "Đỉnh\tBậc vào\tBậc ra\tKhuyên" << endl;
        cout << "----\t-------\t------\t------" << endl;
        
        for (int i = 0; i < numVertices; i++) {
            int inDegree = 0, outDegree = 0;
            
            // Tính bậc ra
            outDegree = adjList[i].size();
            
            // Tính bậc vào
            for (int j = 0; j < numVertices; j++) {
                for (const Edge& edge : adjList[j]) {
                    if (edge.to == i && j != i) {
                        inDegree++;
                    }
                }
            }
            
            cout << i << "\t" << inDegree << "\t" << outDegree 
                 << "\t" << selfLoopCount[i] << endl;
        }
    }
    
    void printGeneralGraph() {
        cout << "Biểu diễn đồ thị tổng quát:" << endl;
        
        for (int i = 0; i < numVertices; i++) {
            cout << "Đỉnh " << i << ":" << endl;
            
            if (selfLoopCount[i] > 0) {
                cout << "  Khuyên: " << selfLoopCount[i] << endl;
            }
            
            cout << "  Kề: ";
            for (const Edge& edge : adjList[i]) {
                if (edge.to != i) {  // Không in khuyên ở đây
                    cout << edge.to << (edge.isDirected ? "(→)" : "(↔)") << " ";
                }
            }
            cout << endl;
        }
        cout << endl;
    }
    
    // Tìm tất cả chu trình trong đồ thị tổng quát
    void findAllCycles() {
        cout << "Tìm tất cả chu trình trong đồ thị:" << endl;
        
        vector<vector<int>> cycles;
        vector<int> currentPath;
        vector<bool> inPath(numVertices, false);
        
        for (int i = 0; i < numVertices; i++) {
            findCyclesFromVertex(i, currentPath, inPath, cycles);
        }
        
        if (cycles.empty()) {
            cout << "Không có chu trình" << endl;
        } else {
            cout << "Tìm thấy " << cycles.size() << " chu trình:" << endl;
            for (int i = 0; i < cycles.size(); i++) {
                cout << "Chu trình " << i + 1 << ": ";
                for (int j = 0; j < cycles[i].size(); j++) {
                    cout << cycles[i][j];
                    if (j < cycles[i].size() - 1) cout << " -> ";
                }
                cout << " -> " << cycles[i][0] << endl;
            }
        }
    }
    
private:
    void findCyclesFromVertex(int start, vector<int>& path, 
                             vector<bool>& inPath, vector<vector<int>>& cycles) {
        path.push_back(start);
        inPath[start] = true;
        
        for (const Edge& edge : adjList[start]) {
            int next = edge.to;
            
            if (next == start) continue;  // Bỏ qua khuyên
            
            if (inPath[next]) {
                // Tìm thấy chu trình
                auto it = find(path.begin(), path.end(), next);
                if (it != path.end()) {
                    vector<int> cycle(it, path.end());
                    cycles.push_back(cycle);
                }
            } else if (find(path.begin(), path.end(), next) == path.end()) {
                findCyclesFromVertex(next, path, inPath, cycles);
            }
        }
        
        path.pop_back();
        inPath[start] = false;
    }
};

// Demo cho đồ thị tổng quát
void demonstrateGeneralGraphDFS() {
    cout << "=== DEMO DFS CHO ĐỒ THỊ TỔNG QUÁT ===" << endl << endl;
    
    GeneralGraphDFS graph(6);
    
    // Tạo đồ thị tổng quát phức tạp
    cout << "Tạo đồ thị tổng quát:" << endl;
    
    // Cạnh vô hướng
    graph.addEdge(0, 1, false);
    graph.addEdge(1, 2, false);
    
    // Cạnh có hướng
    graph.addEdge(2, 3, true);
    graph.addEdge(3, 4, true);
    
    // Cạnh bội
    graph.addEdge(0, 1, false);  // Cạnh bội vô hướng
    graph.addEdge(2, 3, true);   // Cạnh bội có hướng
    
    // Khuyên
    graph.addEdge(4, 4, true);   // Khuyên có hướng
    graph.addEdge(5, 5, false);  // Khuyên vô hướng
    
    // Cạnh tạo chu trình
    graph.addEdge(4, 1, true);   // Chu trình có hướng
    graph.addEdge(3, 0, false);  // Chu trình hỗn hợp
    
    cout << endl;
    graph.printGeneralGraph();
    
    // DFS trên đồ thị tổng quát
    graph.dfsComplete();
    
    // Tìm chu trình
    graph.findAllCycles();
}

int main() {
    demonstrateGeneralGraphDFS();
    return 0;
}